File: timerx-core/src/main/java/timerx/StringBuilderTimeFormatter.java
package timerx;

import static timerx.Constants.TimeValues.MILLIS_IN_SECOND;
import static timerx.Constants.TimeValues.MINUTES_IN_HOUR;
import static timerx.Constants.TimeValues.NONE;
import static timerx.Constants.TimeValues.SECONDS_IN_MINUTE;
import static timerx.TimeUnitType.HOURS;
import static timerx.TimeUnitType.MINUTES;
import static timerx.TimeUnitType.R_MILLISECONDS;
import static timerx.TimeUnitType.SECONDS;

import androidx.annotation.NonNull;

/**
 * Optimized time formatter based on string builder
 */
class StringBuilderTimeFormatter extends TimeFormatter {

  private final Semantic semantic;
  private final StringBuilder mutableString;
  private final TimeContainer timeContainer = new TimeContainer();

  StringBuilderTimeFormatter(@NonNull Semantic semantic) {
    this.semantic = semantic;
    mutableString = new StringBuilder(semantic.strippedFormat.length());
    mutableString.append(semantic.strippedFormat);
  }

  long getOptimalDelay() {
    long delay = 100;
    if (semantic.has(R_MILLISECONDS)) {
      if (semantic.rMillisPosition.length() == 2) {
        delay = 10;
      } else if (semantic.rMillisPosition.length() > 2) {
        delay = 1;
      }
    }
    return delay;
  }

  @NonNull
  String getFormat() {
    return semantic.getFormat();
  }

  @NonNull
  CharSequence format(long millis) {
    TimeContainer units = timeUnitsOf(millis);
    long millisToShow = NONE;
    long secondsToShow = NONE;
    long minutesToShow = NONE;
    long hoursToShow = NONE;
    if (semantic.has(R_MILLISECONDS)) {
      millisToShow = (semantic.has(SECONDS)) ? units.remMillis : units.millis;
    }
    if (semantic.has(SECONDS)) {
      secondsToShow = (semantic.has(MINUTES)) ? units.remSeconds : units.seconds;
    }
    if (semantic.has(MINUTES)) {
      minutesToShow = (semantic.has(HOURS)) ? units.remMinutes : units.minutes;
    }
    if (semantic.has(HOURS)) {
      hoursToShow = units.hours;
    }
    applyFormat(millisToShow, secondsToShow, minutesToShow, hoursToShow);
    return mutableString;
  }

  @NonNull
  private TimeContainer timeUnitsOf(long millis) {
    long seconds = millis / MILLIS_IN_SECOND;
    long minutes = seconds / SECONDS_IN_MINUTE;
    long hours = minutes / MINUTES_IN_HOUR;
    long remMillis = millis % MILLIS_IN_SECOND;
    long remSeconds = seconds - minutes * SECONDS_IN_MINUTE;
    long remMinutes = minutes - hours * MINUTES_IN_HOUR;
    return timeContainer
        .setMillis(millis)
        .setSeconds(seconds)
        .setMinutes(minutes)
        .setHours(hours)
        .setRemMillis(remMillis)
        .setRemSeconds(remSeconds)
        .setRemMinutes(remMinutes);
  }

  private void applyFormat(long millisToShow, long secondsToShow,
      long minutesToShow, long hoursToShow) {
    if (millisToShow != NONE) updateString(millisToShow, R_MILLISECONDS);
    if (secondsToShow != NONE) updateString(secondsToShow, SECONDS);
    if (minutesToShow != NONE) updateString(minutesToShow, MINUTES);
    if (hoursToShow != NONE) updateString(hoursToShow, HOURS);
  }

  private void updateString(long time, TimeUnitType timeUnitType) {
    Position position = positionOfUnit(timeUnitType);
    int timeLength = lengthOf(time);
    if (!semantic.hasOnlyRMillis() && timeUnitType == R_MILLISECONDS) {
      if (semantic.rMillisPosition.length() < 3 && timeLength < 3) {
        int difference = 3 - semantic.rMillisPosition.length();
        time /= Math.pow(10, difference);
      }
      if (semantic.rMillisPosition.length() < timeLength) {
        int difference = timeLength - semantic.rMillisPosition.length();
        time /= Math.pow(10, difference);
      }
    }
    int updatedTimeLength = lengthOf(time);
    int range = Math.max(position.end - position.start, updatedTimeLength - 1);
    for (int i = position.end; i >= position.end - range; i--) {
      char ch = (char) ('0' + (time % 10));
      if (i >= position.start) {
        mutableString.setCharAt(i, ch);
      } else {
        mutableString.insert(Math.max(i + 1, 0), ch);
      }
      time /= 10;
    }
  }

  @NonNull
  private Position positionOfUnit(TimeUnitType timeUnitType) {
    switch (timeUnitType) {
      case HOURS:
        return semantic.hoursPosition;
      case MINUTES:
        return semantic.minutesPosition;
      case SECONDS:
        return semantic.secondsPosition;
      case R_MILLISECONDS:
        return semantic.rMillisPosition;
    }
    throw new IllegalStateException();
  }

  private int lengthOf(long number) {
    int length = 0;
    long temp = 1;
    while (temp <= number) {
      length++;
      temp *= 10;
    }
    return length;
  }
}

File: timerx-core/src/main/java/timerx/StringBuilderTimeFormatter.kt
package timerx

import timerx.Constants.TimeValues
import kotlin.math.pow

/**
 * Optimized time formatter based on string builder
 */
internal class StringBuilderTimeFormatter(private val semantic: Semantic) : TimeFormatter {
  
  private val timeContainer = TimeContainer()
  
  private val mutableString = StringBuilder(semantic.strippedFormat.length).apply {
    append(semantic.strippedFormat)
  }
  
  override val optimalDelay: Long
    get() {
      var delay: Long = 100
      if (semantic.has(TimeUnitType.R_MILLISECONDS)) {
        if (semantic.rMillisPosition.length() == 2) {
          delay = 10
        } else if (semantic.rMillisPosition.length() > 2) {
          delay = 1
        }
      }
      return delay
    }
  
  override val format: String
    get() = semantic.format
  
  override fun format(millis: Long): CharSequence {
    val units = timeUnitsOf(millis)
    var millisToShow = TimeValues.NONE
    var secondsToShow = TimeValues.NONE
    var minutesToShow = TimeValues.NONE
    var hoursToShow = TimeValues.NONE
    if (semantic.has(TimeUnitType.R_MILLISECONDS)) {
      millisToShow = if (semantic.has(TimeUnitType.SECONDS)) units.remMillis else units.millis
    }
    if (semantic.has(TimeUnitType.SECONDS)) {
      secondsToShow = if (semantic.has(TimeUnitType.MINUTES)) units.remSeconds else units.seconds
    }
    if (semantic.has(TimeUnitType.MINUTES)) {
      minutesToShow = if (semantic.has(TimeUnitType.HOURS)) units.remMinutes else units.minutes
    }
    if (semantic.has(TimeUnitType.HOURS)) {
      hoursToShow = units.hours
    }
    applyFormat(millisToShow, secondsToShow, minutesToShow, hoursToShow)
    return mutableString
  }
  
  private fun timeUnitsOf(millis: Long): TimeContainer {
    val seconds = millis / TimeValues.MILLIS_IN_SECOND
    val minutes = seconds / TimeValues.SECONDS_IN_MINUTE
    val hours = minutes / TimeValues.MINUTES_IN_HOUR
    val remMillis = millis % TimeValues.MILLIS_IN_SECOND
    val remSeconds = seconds - minutes * TimeValues.SECONDS_IN_MINUTE
    val remMinutes = minutes - hours * TimeValues.MINUTES_IN_HOUR
    return timeContainer.apply {
      this.millis = millis
      this.seconds = seconds
      this.minutes = minutes
      this.hours = hours
      this.remMillis = remMillis
      this.remSeconds = remSeconds
      this.remMinutes = remMinutes
    }
  }
  
  private fun applyFormat(millisToShow: Long, secondsToShow: Long,
                          minutesToShow: Long, hoursToShow: Long) {
    if (millisToShow != TimeValues.NONE) updateString(millisToShow, TimeUnitType.R_MILLISECONDS)
    if (secondsToShow != TimeValues.NONE) updateString(secondsToShow, TimeUnitType.SECONDS)
    if (minutesToShow != TimeValues.NONE) updateString(minutesToShow, TimeUnitType.MINUTES)
    if (hoursToShow != TimeValues.NONE) updateString(hoursToShow, TimeUnitType.HOURS)
  }
  
  private fun updateString(time: Long, timeUnitType: TimeUnitType) {
    var timeVar = time
    val position = positionOfUnit(timeUnitType)
    val timeLength = lengthOf(timeVar)
    if (!semantic.hasOnlyRMillis() && timeUnitType === TimeUnitType.R_MILLISECONDS) {
      if (semantic.rMillisPosition.length() < 3 && timeLength < 3) {
        val difference = 3 - semantic.rMillisPosition.length()
        timeVar /= 10.0.pow(difference.toDouble()).toLong()
      }
      if (semantic.rMillisPosition.length() < timeLength) {
        val difference = timeLength - semantic.rMillisPosition.length()
        timeVar /= 10.0.pow(difference.toDouble()).toLong()
      }
    }
    val updatedTimeLength = lengthOf(timeVar)
    val range = (position.end - position.start).coerceAtLeast(updatedTimeLength - 1)
    for (i in position.end downTo position.end - range) {
      val ch = ('0'.code.toLong() + timeVar % 10).toInt().toChar()
      if (i >= position.start) {
        mutableString.setCharAt(i, ch)
      } else {
        mutableString.insert((i + 1).coerceAtLeast(0), ch)
      }
      timeVar /= 10
    }
  }
  
  private fun positionOfUnit(timeUnitType: TimeUnitType): Position {
    return when (timeUnitType) {
      TimeUnitType.HOURS -> semantic.hoursPosition
      TimeUnitType.MINUTES -> semantic.minutesPosition
      TimeUnitType.SECONDS -> semantic.secondsPosition
      TimeUnitType.R_MILLISECONDS -> semantic.rMillisPosition
    }
  }
  
  private fun lengthOf(number: Long): Int {
    var length = 0
    var temp: Long = 1
    while (temp <= number) {
      length++
      temp *= 10
    }
    return length
  }
}